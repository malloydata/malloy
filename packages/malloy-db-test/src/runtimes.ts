/*
 * Copyright 2021 Google LLC
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 */

import {
  EmptyURLReader,
  Result,
  MalloyQueryData,
  SingleConnectionRuntime,
} from "@malloydata/malloy";
import { BigQueryConnection } from "@malloydata/db-bigquery";
import { PooledPostgresConnection } from "@malloydata/db-postgres";
import { DuckDBConnection } from "@malloydata/db-duckdb";
import { RunSQLOptions } from "@malloydata/malloy/src/malloy";

export class BigQueryTestConnection extends BigQueryConnection {
  // we probably need a better way to do this.

  public async runSQL(
    sqlCommand: string,
    options?: RunSQLOptions
  ): Promise<MalloyQueryData> {
    try {
      return await super.runSQL(sqlCommand, options);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(`Error in SQL:\n ${sqlCommand}`);
      throw e;
    }
  }
}

export class PostgresTestConnection extends PooledPostgresConnection {
  // we probably need a better way to do this.

  public async runSQL(
    sqlCommand: string,
    options?: RunSQLOptions
  ): Promise<MalloyQueryData> {
    try {
      return await super.runSQL(sqlCommand, options);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(`Error in SQL:\n ${sqlCommand}`);
      throw e;
    }
  }
}

export class DuckDBTestConnection extends DuckDBConnection {
  // we probably need a better way to do this.

  constructor(name: string) {
    super(name);
  }

  public async runSQL(
    sqlCommand: string,
    options?: RunSQLOptions
  ): Promise<MalloyQueryData> {
    try {
      return await super.runSQL(sqlCommand, options);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(`Error in SQL:\n ${sqlCommand}`);
      throw e;
    }
  }
}

const files = new EmptyURLReader();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function rows(qr: Result): any[] {
  return qr.data.value;
}

export const allDatabases = ["postgres", "bigquery", "duckdb"];
type RuntimeDatabaseNames = typeof allDatabases[number];

export class RuntimeList {
  runtimeMap = new Map<string, SingleConnectionRuntime>();

  constructor(databaseList: RuntimeDatabaseNames[] | undefined = undefined) {
    for (const dbName of databaseList || allDatabases) {
      switch (dbName) {
        case "bigquery":
          this.runtimeMap.set(
            "bigquery",
            new SingleConnectionRuntime(
              files,
              new BigQueryTestConnection(
                "bigquery",
                {},
                { defaultProject: "malloy-data" }
              )
            )
          );
          break;
        case "postgres":
          {
            const pg = new PostgresTestConnection("postgres");
            this.runtimeMap.set(
              "postgres",
              new SingleConnectionRuntime(files, pg)
            );
          }
          break;
        case "duckdb": {
          const duckdb = new DuckDBTestConnection("duckdb");
          this.runtimeMap.set(
            "duckdb",
            new SingleConnectionRuntime(files, duckdb)
          );
        }
      }
    }
  }

  async closeAll(): Promise<void> {
    for (const [_key, runtime] of this.runtimeMap) {
      if (runtime.connection.isPool()) runtime.connection.drain();
    }
  }
}
