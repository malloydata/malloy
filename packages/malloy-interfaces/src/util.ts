import {RESERVED_WORDS} from './reserved_words';

function shouldQuoteIdentifier(name: string) {
  const containsFunnyCharacters = !name.match(/^[A-Za-z_][A-Za-z_0-9]*$/);
  const isReserved = RESERVED_WORDS.includes(name.toLowerCase());
  return containsFunnyCharacters || isReserved;
}

export function maybeQuoteIdentifier(name: string): string {
  const path = name.split('.');
  for (let i = 0; i < path.length; i++) {
    if (shouldQuoteIdentifier(path[i])) {
      path[i] = `\`${path[i]}\``;
    }
  }
  return path.join('.');
}
