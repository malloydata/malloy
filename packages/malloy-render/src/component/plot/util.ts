import {Explore, ExploreField, Field} from '@malloydata/malloy';

export function walkFields(e: Explore, cb: (f: Field) => void) {
  e.allFields.forEach(f => {
    cb(f);
    if (f.isExplore()) {
      walkFields(f, cb);
    }
  });
}

export function getFieldPathArrayFromRoot(f: Field | Explore) {
  const paths = f.isExplore() && !f.isExploreField() ? [] : [f.name];
  let parent = f.parentExplore;
  while (parent?.isExploreField()) {
    paths.unshift(parent.name);
    parent = parent.parentExplore;
  }
  return paths;
}

export function getFieldPathFromRoot(f: Field | Explore) {
  return getFieldPathArrayFromRoot(f).join('.');
}

export function getFieldPathBetweenFields(
  parentField: Field | Explore,
  childField: Field | Explore
): string {
  const parentPath = getFieldPathArrayFromRoot(parentField);
  const childPath = getFieldPathArrayFromRoot(childField);
  const startIndex = parentPath.length;

  let i = 0;
  while (parentPath[i]) {
    if (parentPath[i] !== childPath[i])
      throw new Error(
        'Tried to get path from parent field to child field, but parent field is not a parent of child field.'
      );
    i++;
  }

  return childPath.slice(startIndex).join('.');
}

export function getFieldFromRootPath(root: Explore, path: string): Field {
  const pathParts = path.split('.');
  let curr: Field | ExploreField | Explore = root;
  for (const part of pathParts) {
    if (curr.isExplore()) {
      curr = curr.allFields.find(f => f.name === part)!;
    } else {
      throw new Error('Tried to get field from path, but path is invalid');
    }
  }
  if (curr.isExplore() && !curr.isExploreField())
    throw new Error('Tried to get field from path, but got root Explore');
  return curr;
}
