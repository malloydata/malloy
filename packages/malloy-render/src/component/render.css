:host {
  --malloy-theme--table-row-height: 28px;
  --malloy-theme--table-font-size: 12px;
  --malloy-theme--table-header-color: #5d626b;
  --malloy-theme--table-header-weight: bold;
  --malloy-theme--table-body-color: #727883;
  --malloy-theme--table-body-weight: 400;
  --malloy-theme--table-border: 1px solid #e5e7eb;
  --malloy-theme--table-background: white;
  --malloy-theme--table-gutter-size: 8px;
  --malloy-theme--table-pinned-background: #f5fafc;
  --malloy-theme--table-pinned-border: 1px solid #daedf3;
  --malloy-theme--font-family: Inter, system-ui, sans-serif;

  font-family: var(--malloy-render--font-family);
  font-size: var(--malloy-render--table-font-size);
}

@supports (font-variation-settings: normal) {
  :host {
    font-family:
      InterVariable,
      Inter,
      system-ui sans-serif;

    font-feature-settings:
      'liga' 1,
      'calt' 1;
  }
}
