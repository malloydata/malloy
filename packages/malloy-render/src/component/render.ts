/*
 * Copyright 2023 Google LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files
 * (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {ModelDef, QueryResult, Result, Tag} from '@malloydata/malloy';
import {LitElement, html, css, PropertyValues} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';
import './table';
import './bar-chart';
import {provide} from '@lit/context';
import {resultContext} from './result-context';
import {
  RenderResultMetadata,
  getResultMetadata,
} from './render-result-metadata';

// Get the first valid theme value or fallback to CSS variable
function getThemeValue(prop: string, ...themes: Array<Tag | undefined>) {
  let value: string | undefined;
  for (const theme of themes) {
    value = theme?.text(prop);
    if (typeof value !== 'undefined') break;
  }
  // If no theme overrides, convert prop name from camelCase to kebab and pull from --malloy-theme-- variable
  return (
    value ??
    `var(--malloy-theme--${prop
      .replace(/([a-z])([A-Z])/g, '$1-$2')
      .toLowerCase()})`
  );
}

@customElement('malloy-render')
export class MalloyRender extends LitElement {
  static override styles = css`
    :host {
      --malloy-theme--table-row-height: 28px;
      --malloy-theme--table-font-size: 12px;
      --malloy-theme--table-header-color: #5d626b;
      --malloy-theme--table-header-weight: bold;
      --malloy-theme--table-body-color: #727883;
      --malloy-theme--table-body-weight: 400;
      --malloy-theme--table-border: 1px solid #e5e7eb;
      --malloy-theme--table-background: white;
      --malloy-theme--table-gutter-size: 15px;
      --malloy-theme--table-pinned-background: #f5fafc;
      --malloy-theme--table-pinned-border: 1px solid #daedf3;
      --malloy-theme--font-family: Inter, system-ui, sans-serif;

      font-family: var(--malloy-render--font-family);
      font-size: var(--malloy-render--table-font-size);
    }

    @supports (font-variation-settings: normal) {
      :host {
        font-family:
          InterVariable,
          Inter,
          system-ui sans-serif;

        font-feature-settings:
          'liga' 1,
          'calt' 1;
      }
    }
  `;

  @property({attribute: false})
  result?: Result;

  @property({attribute: false})
  queryResult?: QueryResult;

  @property({attribute: false})
  modelDef?: ModelDef;

  @state()
  private _result!: Result;

  @provide({context: resultContext})
  metadata!: RenderResultMetadata;

  willUpdate(changedProperties: PropertyValues<this>) {
    if (
      changedProperties.has('result') ||
      changedProperties.has('queryResult') ||
      changedProperties.has('modelDef')
    ) {
      if (this.result) this._result = this.result;
      else if (this.queryResult && this.modelDef) {
        this._result = new Result(this.queryResult, this.modelDef);
      }
      this.metadata = getResultMetadata(this._result);
      const modelTag = this._result.modelTag;
      const {tag: resultTag} = this._result.tagParse();
      const modelTheme = modelTag.tag('theme');
      const localTheme = resultTag.tag('theme');
      this.updateTheme(modelTheme, localTheme);
    }
  }

  updateTheme(modelTheme?: Tag, localTheme?: Tag) {
    const tableRowHeight = getThemeValue(
      'tableRowHeight',
      localTheme,
      modelTheme
    );
    const tableBodyColor = getThemeValue(
      'tableBodyColor',
      localTheme,
      modelTheme
    );
    const tableFontSize = getThemeValue(
      'tableFontSize',
      localTheme,
      modelTheme
    );
    const tableHeaderColor = getThemeValue(
      'tableHeaderColor',
      localTheme,
      modelTheme
    );
    const tableHeaderWeight = getThemeValue(
      'tableHeaderWeight',
      localTheme,
      modelTheme
    );
    const tableBodyWeight = getThemeValue(
      'tableBodyWeight',
      localTheme,
      modelTheme
    );
    const tableBorder = getThemeValue('tableBorder', localTheme, modelTheme);
    const tableBackground = getThemeValue(
      'tableBackground',
      localTheme,
      modelTheme
    );
    const tableGutterSize = getThemeValue(
      'tableGutterSize',
      localTheme,
      modelTheme
    );
    const tablePinnedBackground = getThemeValue(
      'tablePinnedBackground',
      localTheme,
      modelTheme
    );
    const tablePinnedBorder = getThemeValue(
      'tablePinnedBorder',
      localTheme,
      modelTheme
    );
    const fontFamily = getThemeValue('fontFamily', localTheme, modelTheme);

    this.style.setProperty('--malloy-render--table-row-height', tableRowHeight);
    this.style.setProperty('--malloy-render--table-body-color', tableBodyColor);
    this.style.setProperty('--malloy-render--table-font-size', tableFontSize);
    this.style.setProperty('--malloy-render--font-family', fontFamily);
    this.style.setProperty(
      '--malloy-render--table-header-color',
      tableHeaderColor
    );
    this.style.setProperty(
      '--malloy-render--table-header-weight',
      tableHeaderWeight
    );
    this.style.setProperty(
      '--malloy-render--table-body-weight',
      tableBodyWeight
    );
    this.style.setProperty('--malloy-render--table-border', tableBorder);
    this.style.setProperty(
      '--malloy-render--table-background',
      tableBackground
    );
    this.style.setProperty(
      '--malloy-render--table-gutter-size',
      tableGutterSize
    );
    this.style.setProperty(
      '--malloy-render--table-pinned-background',
      tablePinnedBackground
    );
    this.style.setProperty(
      '--malloy-render--table-pinned-border',
      tablePinnedBorder
    );
  }

  override render() {
    return html`<malloy-table
      exportparts="table-container: container"
      .data=${this._result.data}
    ></malloy-table>`;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'malloy-render': MalloyRender;
  }
}
