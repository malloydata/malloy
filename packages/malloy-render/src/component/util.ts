/*
 * Copyright 2023 Google LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files
 * (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
import {Explore, Field} from '@malloydata/malloy';

function getLocationInParent(f: Field | Explore) {
  const parent = f.parentExplore;
  return parent?.allFields.findIndex(pf => pf.name === f.name) ?? -1;
}

export function isLastChild(f: Field | Explore) {
  if (f.parentExplore)
    return getLocationInParent(f) === f.parentExplore.allFields.length - 1;
  return true;
}

export function isFirstChild(f: Field | Explore) {
  return getLocationInParent(f) === 0;
}

export function valueIsNumber(f: Field, v: unknown): v is number {
  return f.isAtomicField() && f.isNumber() && v !== null;
}

export function valueIsString(f: Field, s: unknown): s is string {
  return f.isAtomicField() && f.isString() && s !== null;
}

export function valueIsDateTime(f: Field, v: unknown): v is Date {
  return f.isAtomicField() && (f.isDate() || f.isTimestamp()) && v !== null;
}

export function getTextWidth(
  text: string,
  font: string,
  canvasToUse?: HTMLCanvasElement
) {
  const canvas = canvasToUse ?? document.createElement('canvas');
  const context = canvas.getContext('2d')!;
  context.font = font;
  const metrics = context.measureText(text);
  return metrics.width;
}

export function clamp(s: number, e: number, v: number) {
  return Math.max(s, Math.min(e, v));
}

export function getFieldKey(f: Field | Explore) {
  return JSON.stringify(f.fieldPath);
}

export function getRangeSize(range: [number, number]) {
  return range[1] - range[0] + 1;
}
