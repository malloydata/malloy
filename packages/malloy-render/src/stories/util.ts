import {LoaderFunction, Args} from '@storybook/types';
import {HtmlRenderer} from '@storybook/html';
import {SingleConnectionRuntime} from '@malloydata/malloy';
import {DuckDBWASMConnection} from '@malloydata/db-duckdb/wasm';
import {HTMLView} from '../html';
import {RendererOptions} from '../renderer_types';

export type QueryOptions = {
  script: string;
  source: string;
  view: string;
  connection: Promise<DuckDBWASMConnection>;
};

export function createLoader(
  script: string
): LoaderFunction<HtmlRenderer, Args> {
  return async context => {
    return {
      result: await runQuery({
        script,
        source: context.args['source'],
        view: context.args['view'],
        connection: context.globals['getConnection'](),
      }),
    };
  };
}

export async function loadModel({
  script,
  connection,
}: {
  script: string;
  connection: Promise<DuckDBWASMConnection>;
}) {
  const conn = await connection;
  return new SingleConnectionRuntime(conn).loadModel(script);
}

export async function runQuery({
  script,
  source,
  view,
  connection,
}: QueryOptions) {
  const model = await loadModel({script, connection});
  const runner = model.loadQuery(`run: ${source} -> ${view}`);
  const result = await runner.run();
  return result;
}

/* Legacy Renderer */

type RenderOptions = QueryOptions & {
  classes?: '';
};

async function runAndRender(
  {script, source, view, connection}: RenderOptions,
  options: RendererOptions = {dataStyles: {}}
) {
  const viewer = new HTMLView(document);
  const result = await runQuery({script, source, view, connection});
  return await viewer.render(result, options);
}

export function renderMalloyLegacy(options: RenderOptions) {
  const div = document.createElement('div');
  runAndRender(options, {
    dataStyles: {},
    isDrillingEnabled: true,
    onDrill: (
      drillQuery: string,
      _target: HTMLElement,
      _drillFilters: string[]
    ) => {
      navigator.clipboard.writeText(drillQuery);
    },
  }).then(el => {
    if (options.classes) el.classList.add(options.classes);
    div.replaceChildren(el);
  });
  return div;
}
