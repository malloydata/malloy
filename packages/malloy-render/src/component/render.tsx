/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 *  LICENSE file in the root directory of this source tree.
 */

import {
  ExploreField,
  ModelDef,
  QueryResult,
  Result,
  Tag,
} from '@malloydata/malloy';
import {
  Show,
  createContext,
  createEffect,
  createMemo,
  useContext,
} from 'solid-js';
import {getResultMetadata} from './render-result-metadata';
import {ResultContext} from './result-context';
import './render.css';
import {ComponentOptions, ICustomElement} from 'component-register';
import {applyRenderer} from './apply-renderer';
import {MalloyClickEventPayload, VegaConfigHandler} from './types';

export type MalloyRenderProps = {
  result?: Result;
  queryResult?: QueryResult;
  modelDef?: ModelDef;
  scrollEl?: HTMLElement;
  onClick?: (payload: MalloyClickEventPayload) => void;
  vegaConfigOverride?: VegaConfigHandler;
};

const ConfigContext = createContext<{
  onClick?: (payload: MalloyClickEventPayload) => void;
  vegaConfigOverride?: VegaConfigHandler;
}>();

export const useConfig = () => {
  const config = useContext(ConfigContext);
  if (!config)
    throw new Error(
      'ConfigContext missing a value; did you provide a ConfigProvider?'
    );
  return config;
};

export function MalloyRender(
  props: MalloyRenderProps,
  {element}: ComponentOptions
) {
  const result = createMemo(() => {
    if (props.result) return props.result;
    else if (props.queryResult && props.modelDef)
      return new Result(props.queryResult, props.modelDef);
    else return null;
  });

  return (
    <Show when={result()}>
      <ConfigContext.Provider
        value={{
          onClick: props.onClick,
          vegaConfigOverride: props.vegaConfigOverride,
        }}
      >
        <MalloyRenderInner
          result={result()!}
          element={element}
          scrollEl={props.scrollEl}
          vegaConfigOverride={props.vegaConfigOverride}
        />
      </ConfigContext.Provider>
    </Show>
  );
}

export function MalloyRenderInner(props: {
  result: Result;
  element: ICustomElement;
  scrollEl?: HTMLElement;
  vegaConfigOverride?: VegaConfigHandler;
}) {
  const metadata = createMemo(() =>
    getResultMetadata(props.result, {
      getVegaConfigOverride: props.vegaConfigOverride,
    })
  );
  const tags = () => {
    const modelTag = metadata().modelTag;
    const resultTag = metadata().resultTag;
    const modelTheme = modelTag.tag('theme');
    const localTheme = resultTag.tag('theme');
    return {
      modelTag,
      resultTag,
      modelTheme,
      localTheme,
    };
  };

  createEffect(() => {
    if (props.element) {
      const style = generateThemeStyle(tags().modelTheme, tags().localTheme);
      for (const [key, value] of Object.entries(style)) {
        props.element['style'].setProperty(key, value);
      }
    }
  });

  const rendering = () => {
    return applyRenderer({
      // TODO: figure out what to do about the diffs between top level Explore vs. ExploreFields/AtomicFields
      field: props.result.resultExplore as ExploreField,
      dataColumn: props.result.data,
      resultMetadata: metadata(),
      tag: tags().resultTag,
      customProps: {
        table: {
          scrollEl: props.scrollEl,
        },
        dashboard: {
          scrollEl: props.scrollEl,
        },
      },
    });
  };

  return (
    <ResultContext.Provider value={metadata()}>
      {rendering().renderValue}
    </ResultContext.Provider>
  );
}

// Get the first valid theme value or fallback to CSS variable
function getThemeValue(prop: string, ...themes: Array<Tag | undefined>) {
  let value: string | undefined;
  for (const theme of themes) {
    value = theme?.text(prop);
    if (typeof value !== 'undefined') break;
  }
  // If no theme overrides, convert prop name from camelCase to kebab and pull from --malloy-theme-- variable
  return (
    value ??
    `var(--malloy-theme--${prop
      .replace(/([a-z])([A-Z])/g, '$1-$2')
      .toLowerCase()})`
  );
}

function generateThemeStyle(modelTheme?: Tag, localTheme?: Tag) {
  const style: Record<string, string> = {};

  const tableRowHeight = getThemeValue(
    'tableRowHeight',
    localTheme,
    modelTheme
  );
  const tableBodyColor = getThemeValue(
    'tableBodyColor',
    localTheme,
    modelTheme
  );
  const tableFontSize = getThemeValue('tableFontSize', localTheme, modelTheme);
  const tableHeaderColor = getThemeValue(
    'tableHeaderColor',
    localTheme,
    modelTheme
  );
  const tableHeaderWeight = getThemeValue(
    'tableHeaderWeight',
    localTheme,
    modelTheme
  );
  const tableBodyWeight = getThemeValue(
    'tableBodyWeight',
    localTheme,
    modelTheme
  );
  const tableBorder = getThemeValue('tableBorder', localTheme, modelTheme);
  const tableBackground = getThemeValue(
    'tableBackground',
    localTheme,
    modelTheme
  );
  const tableGutterSize = getThemeValue(
    'tableGutterSize',
    localTheme,
    modelTheme
  );
  const tablePinnedBackground = getThemeValue(
    'tablePinnedBackground',
    localTheme,
    modelTheme
  );
  const tablePinnedBorder = getThemeValue(
    'tablePinnedBorder',
    localTheme,
    modelTheme
  );
  const fontFamily = getThemeValue('fontFamily', localTheme, modelTheme);

  style['--malloy-render--table-row-height'] = tableRowHeight;
  style['--malloy-render--table-body-color'] = tableBodyColor;
  style['--malloy-render--table-font-size'] = tableFontSize;
  style['--malloy-render--font-family'] = fontFamily;
  style['--malloy-render--table-header-color'] = tableHeaderColor;
  style['--malloy-render--table-header-weight'] = tableHeaderWeight;
  style['--malloy-render--table-body-weight'] = tableBodyWeight;
  style['--malloy-render--table-border'] = tableBorder;
  style['--malloy-render--table-background'] = tableBackground;
  style['--malloy-render--table-gutter-size'] = tableGutterSize;
  style['--malloy-render--table-pinned-background'] = tablePinnedBackground;
  style['--malloy-render--table-pinned-border'] = tablePinnedBorder;
  return style;
}
