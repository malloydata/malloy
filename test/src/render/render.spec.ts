/*
 * Copyright 2023 Google LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files
 * (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {RuntimeList} from '../runtimes';
import {describeIfDatabaseAvailable} from '../util';
import {HTMLView} from '@malloydata/render';
import {JSDOM} from 'jsdom';

async function runUnsupportedRenderTest(
  connectionName: string,
  runtimes: RuntimeList,
  expr: string,
  rendered: string
) {
  const runtime = runtimes.runtimeMap.get(connectionName);
  expect(runtime).toBeDefined();
  if (runtime) {
    const src = `
      sql: sql_block is {
        select: """SELECT ${expr} AS test"""
        connection: "${connectionName}"
      }
      query: q is from_sql(sql_block)->{ project: *}
    `;
    const result = await runtime.loadModel(src).loadQueryByName('q').run();
    // console.log("DATA", result.data.toObject());
    const document = new JSDOM().window.document;
    const html = await new HTMLView(document).render(result.data, {
      dataStyles: {},
    });
    expect(html.innerHTML).toContain('<thead>');
    expect(html.innerHTML).toContain(rendered);
    // console.log(html.innerHTML);
  }
}

const [describe, databases] = describeIfDatabaseAvailable([
  'bigquery',
  'postgres',
  'duckdb',
]);
describe('rendering results', () => {
  const runtimes = new RuntimeList(databases);

  afterAll(async () => {
    await runtimes.closeAll();
  });

  test('can render table', async () => {
    const runtime = runtimes.runtimeMap.get('bigquery');
    expect(runtime).toBeDefined();
    if (runtime) {
      const src = `
        query: table('malloy-data.faa.flights') -> {
          group_by: carrier
          aggregate: flight_count is count()
        }
      `;
      const result = await runtime.loadQuery(src).run();
      const document = new JSDOM().window.document;
      await new HTMLView(document).render(result.data, {
        dataStyles: {},
      });
    }
  });

  test('can render unsupported bigquery geo types', async () => {
    await runUnsupportedRenderTest(
      'bigquery',
      runtimes,
      "ST_GEOGFROMTEXT('LINESTRING(1 2, 3 4)')",
      'LINESTRING(1 2, 3 4)'
    );
  });

  test('can render unsupported bigquery ip types', async () => {
    await runUnsupportedRenderTest(
      'bigquery',
      runtimes,
      "NET.IP_FROM_STRING('192.168.1.1')",
      '{"type":"Buffer","data":[192,168,1,1]}'
    );
  });

  test('can render unsupported bigquery interval types', async () => {
    await runUnsupportedRenderTest(
      'bigquery',
      runtimes,
      'INTERVAL 1 YEAR',
      '1-0 0 0:0:0'
    );
  });

  test('can render unsupported bigquery time types', async () => {
    await runUnsupportedRenderTest(
      'bigquery',
      runtimes,
      'TIME(10, 10, 1)',
      '10:10:01'
    );
  });

  test('can render unsupported postgres interval types', async () => {
    await runUnsupportedRenderTest(
      'postgres',
      runtimes,
      'make_interval(days => 12)',
      '12 days'
    );
  });

  test('can render unsupported postgres uuid types', async () => {
    await runUnsupportedRenderTest(
      'postgres',
      runtimes,
      "CAST('a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11' AS UUID)",
      'a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11'
    );
  });

  test('can render unsupported postgres inet types', async () => {
    await runUnsupportedRenderTest(
      'postgres',
      runtimes,
      "'192.168.1.1'::inet",
      '192.168.1.1'
    );
  });

  test('can render unsupported postgres macaddr types', async () => {
    await runUnsupportedRenderTest(
      'postgres',
      runtimes,
      "'00:04:E2:36:95:C0'::macaddr",
      '00:04:e2:36:95:c0'
    );
  });

  test('can render supported postgres types', async () => {
    await runUnsupportedRenderTest('postgres', runtimes, '12345', '12,345');
  });

  test('can render supported duckdb types', async () => {
    await runUnsupportedRenderTest('duckdb', runtimes, '12345', '12,345');
  });

  test('can render unsupported duckdb blob types', async () => {
    await runUnsupportedRenderTest(
      'duckdb',
      runtimes,
      "'\\xAA'::BLOB",
      '{"type":"Buffer","data":[170]}'
    );
  });

  test('can render unsupported duckdb uuid types', async () => {
    await runUnsupportedRenderTest(
      'duckdb',
      runtimes,
      "'a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11'::UUID",
      'a0eebc99-9c0b-4ef8-bb6d-6bb9bd380a11'
    );
  });

  test('can render null unsupported types', async () => {
    await runUnsupportedRenderTest(
      'bigquery',
      runtimes,
      'CAST(NULL AS GEOGRAPHY)',
      '<span class="value-null">âˆ…</span>'
    );
  });

  describe('date renderer', () => {
    test('date with timezone rendered correctly', async () => {
      const connectionName = 'duckdb';
      const runtime = runtimes.runtimeMap.get(connectionName);
      expect(runtime).toBeDefined();
      const src = `sql: one is { connection: "${connectionName}"  select: """SELECT 1"""}
      query: mex_query is from_sql(one) -> {
          timezone: 'America/Mexico_City'
          project: mex_time is @2021-02-24 03:05:06
        }
      `;
      const result = await (
        await runtime!.loadModel(src).loadQueryByName('mex_query')
      ).run();
      const document = new JSDOM().window.document;
      const html = await new HTMLView(document).render(result.data, {
        dataStyles: {},
      });

      expect(html).toMatchSnapshot();
    });
  });
});
