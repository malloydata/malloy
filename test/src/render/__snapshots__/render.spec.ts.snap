// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering results date renderer date with timezone rendered correctly 1`] = `
<table
  style="vertical-align: top; border-collapse: collapse; width: 100%;"
>
  <thead>
    <tr>
      <th
        style="padding: 8px; text-align: left;"
      >
        mex_â€‹time (America/Mexico_City)
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style="padding: 8px; vertical-align: top;"
      >
        <span>
          2021-02-24 03:05:06
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
